Extending the Ray Serve integration to allow attributes for Serve deployments

zoltan-fedor opened the pull request:
**Related Issue(s)**: closes #2917

We should be able to set Ray Serve attributes for the nodes of pipelines, like amount of GPU to use, max_concurrent_queries, etc.

Now this is possible from the pipeline yaml file for each node of the pipeline.

**Proposed changes**:
- Extending the Ray Serve integration to allow attributes to be provided for the Serve deployments. Attributes like GPU usage (incl partial GPU usage), `num_concurrent_queries` and so on. See full list of attributes for the `ray.serve.deployment()` method in the [Ray Serve API documentation](https://docs.ray.io/en/latest/serve/package-ref.html)

## Pre-flight checklist
- [x]  I have read the [contributors guidelines](https://github.com/deepset-ai/haystack/blob/master/CONTRIBUTING.md)
- [x] I have [enabled actions on my fork](https://github.com/deepset-ai/haystack/blob/master/CONTRIBUTING.md#forks)
- [x] If this is a code change, I added tests or updated existing ones 
- [x] If this is a code change, I updated the docstrings

zoltan-fedor commented:
Sorry, the schema-generation test keeps failing, even while I have regenerated the schemas in Python 3.8, then in Python 3.7, so not quite sure why they schema generation in the CI are not matching the one I have generated locally.

I hope it won't stop us from merging.

ZanSara commented:
Hey @zoltan-fedor! Thank you for this PR. I will review it shortly :slightly_smiling_face:

ZanSara reviewed (changes requested):
There are a couple of catches here.

The core issue is that you added the `deployment_kwargs` block to the general pipeline schema, without restricting it to Ray pipelines. This needs two actions to be fixed:

1. We need a pipeline test that verifies that Ray attributes cannot be added to base pipelines. It could be something as simple as adding the following test in `tests/pipelines/test_pipeline.yaml.py`, approx line 996:

```python
def test_load_yaml_ray_args_in_pipeline(tmp_path):
    with pytest.raises(PipelineConfigError) as e:
      pipeline = Pipeline.load_from_yaml(
          SAMPLES_PATH / "pipeline" / "ray.haystack-pipeline.yml", pipeline_name="ray_query_pipeline",
      )
```

3. We need to restrict this argument to Ray pipelines. This is rather easy, you just need to add `deployment_kwargs` alongside `replicas` in this part of the schema: https://github.com/deepset-ai/haystack/blob/b78db1cbaf9eafd84781480e49a2c1d6afef5b85/haystack/nodes/_json_schema.py#L312-L321

I hope it's all clear! Let me know if there's any question :slightly_smiling_face: 

----

One last thing: how about renaming the new param to `serve_deployment_kwargs`? Just to clarify it's related to Ray Serve. Otherwise `deployment_kwargs` sounds very generic.

dmigo commented on haystack/pipelines/ray.py:
```diff
@@ -182,16 +190,29 @@ def load_from_yaml(  # type: ignore
+                                  under the `ray.serve.deployment()` method
         """
-        RayDeployment = serve.deployment(_RayDeploymentWrapper, name=component_name, num_replicas=replicas)  # type: ignore
+        RayDeployment = serve.deployment(
+            _RayDeploymentWrapper, name=component_name, num_replicas=replicas, **deployment_kwargs  # type: ignore
+        )
```
Does it make sense to take `num_replicas` from the `deployment_kwargs` as well?
This is a breaking change, but would make more sense to me.

dmigo replied:
@ZanSara

zoltan-fedor replied:
Hi @dmigo,
I agree that it would make more sense to take the `num_replicas` from the `deployment_kwargs` as well, I just did not want to introduce breaking changes. But if you and @ZanSara are okey with that, then I can add it.

ZanSara replied:
As an alternative we could simply unpack all arguments in `deployment_kwargs` and put them at the same level as `replicas`. That would prevent the breaking change.

In any case I don't think a breaking change here is an issue. Feel free to move `replicas` into `deployment_kwargs` as well if it looks better to you. In that case just mind to add the breaking-change label so we won't miss this in the release notes :slightly_smiling_face:

zoltan-fedor replied:
I have made the `replicas` change by moving it below the new `serve_deployment_kwargs`.

zoltan-fedor commented on haystack/json-schemas/haystack-pipeline-1.16.schema.json:
```diff
@@ -0,0 +1,4957 @@
+{
```
Probably a bug in the JSON schema generator. I just follow the instructions and ran the `$ python .github/utils/generate_json_schema.py`, which generated this file and added the reference to the `haystack/json-schemas/haystack-pipeline.schema.json` file, so I left that in, seeing that the change was made by the generator.
I can remove it manually then.

zoltan-fedor commented on haystack/json-schemas/haystack-pipeline.schema.json:
```diff
@@ -158,6 +158,20 @@
           "$ref": "https://raw.githubusercontent.com/deepset-ai/haystack/master/haystack/json-schemas/haystack-pipeline-1.6.0.schema.json"
         }
       ]
+    },
+    {
+      "allOf": [
+        {
+          "properties": {
+            "version": {
+              "const": "1.16"
+            }
+          }
+        },
+        {
+          "$ref": "https://raw.githubusercontent.com/deepset-ai/haystack/master/haystack/json-schemas/haystack-pipeline-1.16.schema.json"
+        }
+      ]
```
Removed.

zoltan-fedor commented:
HI @ZanSara ,
As requested, I have renamed `deployment_kwargs` to `serve_deployment_kwargs`.

I have also added `serve_deployment_kwargs` to :
```
"oneOf": [ 
     { 
         "not": {"required": ["extras"]}, 
         "properties": { 
             "pipelines": { 
                 "title": "Pipelines", 
                 "items": {"properties": {"nodes": {"items": {"not": {"required": ["replicas", "serve_deployment_kwargs"]}}}}}, 
             } 
         }, 
     }, 
```

But that will not make the exception tested in the test to be thrown, so this test will fail:
```
def test_load_yaml_ray_args_in_pipeline(tmp_path):
    with pytest.raises(PipelineConfigError) as e:
      pipeline = Pipeline.load_from_yaml(
          SAMPLES_PATH / "pipeline" / "ray.haystack-pipeline.yml", pipeline_name="ray_query_pipeline",
      )
```

Any recommendations?

ZanSara commented:
You're right about the test! We should add a check in the pipeline loading for the `extras: ray`. I'm going to add a comment about where I'd like to see this check added, hold on

zoltan-fedor commented:
Let me finish the removal of the `replicas`

zoltan-fedor commented:
I have made the `replicas` change by moving it below the new `serve_deployment_kwargs`, but I don't think I am allowed to add the `breaking` label to this PR. Would you mind adding that?

ZanSara commented:
> Let me finish the removal of the replicas

Sorry! I haven't seen this. I'm done anyway, I'll be back in a while to see how the tests look like.

zoltan-fedor commented:
Hi @ZanSara,
The test failures seem to be independent from the changes we have made:
```
requests.exceptions.HTTPError: 502 Server Error: Bad Gateway for url: https://huggingface.co/vblagoje/dpr-question_encoder-single-lfqa-wiki/resolve/main/tokenizer.json
```

I saw you reran the tests but that same issue emerged again. I assume some kind of networking issue - hopefully it will go away and the next run of the test will succeed.
Let me know if you need anything from me.

zoltan-fedor commented:
> There are a couple of catches here.
> 
> The core issue is that you added the `deployment_kwargs` block to the general pipeline schema, without restricting it to Ray pipelines. This needs two actions to be fixed:
> 
>     1. We need a pipeline test that verifies that Ray attributes cannot be added to base pipelines. It could be something as simple as adding the following test in `tests/pipelines/test_pipeline.yaml.py`, approx line 996:
> 
> 
> ```python
> def test_load_yaml_ray_args_in_pipeline(tmp_path):
>     with pytest.raises(PipelineConfigError) as e:
>       pipeline = Pipeline.load_from_yaml(
>           SAMPLES_PATH / "pipeline" / "ray.haystack-pipeline.yml", pipeline_name="ray_query_pipeline",
>       )
> ```
> 
>     3. We need to restrict this argument to Ray pipelines. This is rather easy, you just need to add `deployment_kwargs` alongside `replicas` in this part of the schema: https://github.com/deepset-ai/haystack/blob/b78db1cbaf9eafd84781480e49a2c1d6afef5b85/haystack/nodes/_json_schema.py#L312-L321
> 
> 
> I hope it's all clear! Let me know if there's any question slightly_smiling_face
> 
> One last thing: how about renaming the new param to `serve_deployment_kwargs`? Just to clarify it's related to Ray Serve. Otherwise `deployment_kwargs` sounds very generic.

This all has been done, but unfortunately there is no way to set this "resolved" in GitHub, so I can't get this auto-merged, as GitHub still says "Merging can be performed automatically once the requested changes are addressed.". Basically changes are addressed, but no way to mark them as so.
I am afraid a new review is required to have it auto-merge.

ZanSara reviewed (approved):
Great work, thanks for it! :blush: I have no remarks.
